syntax = "proto3";

package data_types;

message CameraFrame {
    bytes image_data = 1;
    unit32 width = 2;
    unit32 height = 3;
    int64 timestamp_ms = 4;
    string format = 5;
}

message BoundingBox {
    unit32 x_min = 1;
    unit32 y_min = 2;
    unit32 x_max = 3;
    unit32 y_max = 4;
    string label = 5;
    float confidence = 6;
}

message DetectionResult {
    int64 frame_timestamp_ms = 1;
    repeated BoundingBox detections = 2;
}

message DeviceStatus {
    enum ConnectionState {
        DISCONNECTED = 0;
        BLE_PARING = 1;
        WLAN_CONNECTED = 2;
        GRPC_READY = 3;
    }

    ConnectionState state = 1;
    float cpu_temperature_c = 2;
    unit32 frame_rate_fps = 3;
    unit32 processing_latency_ms = 4;
    string firmware_version = 5;
    bool is_sleeping = 6;
}

message Command {
    oneof command_type {
        ConfigRequest config_request = 1;
        ControlAction control_action = 2;
        Heartbeat heartbeat = 3;
    }
}

message ConfigRequest {
    optional float detection_threshold = 1;
    optional unit32 sleep_timeout_sec = 2;
}

message ControlAction {
    enum ActionType {
        REBOOT = 0;
        START_STREAMING = 1;
        STOP_STREAMING = 2;
    }
    ActionType action = 1;
}

message Heartbeat {
    int64 timestamp_ms = 1;
}

message ServerMessage {
    oneof message_type {
        CameraFrame camera_frame = 1;
        DetectionResult detection_result = 2;
        DeviceStatus device_status = 3;
        ConfigResponse config_response = 4;
    }
}

message ClientMessage {
    oneof message_type {
        DeviceStatus device_status = 1;
        ConfigResponse config_response = 2;
    }
}

message ConfigResponse {
    bool success = 1;
    string message = 2;
}